// gauge-ts/dist/helpers/CodeHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const typescript_1 = require("typescript");
class CodeHelper {
    constructor() {
        this.printer = typescript_1.createPrinter();
    }
    getStepTexts(node) {
        let dec = node.decorators;
        let stepDecExp = dec.filter(this.isStepDecorator)[0].expression;
        let arg = stepDecExp.arguments[0];
        if (!arg.text && arg.elements) {
            return arg.elements.map((e) => { return e.text; });
        }
        return [arg.text];
    }
    isStepDecorator(d) {
        let decExp = d.expression;
        return decExp.expression.escapedText === 'Step';
    }
    hasStepDecorator(node) {
        return !!node.decorators && node.decorators.some(this.isStepDecorator);
    }
    hasStepText(node, stepText) {
        let dec = node.decorators;
        let stepDecExp = dec.filter(this.isStepDecorator)[0].expression;
        let arg = stepDecExp.arguments[0];
        if (!arg.text && arg.elements)
            return arg.elements.some((e) => { return e.text === stepText; });
        return arg.text === stepText;
    }
}
exports.CodeHelper = CodeHelper;
