// gauge-ts/dist/processors/DataStoreInitProcessor.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const messages_1 = require("../gen/messages");
const DataStoreFactory_1 = require("../stores/DataStoreFactory");
class DataStoreInitProcessor {
    process(message) {
        return __awaiter(this, void 0, void 0, function* () {
            switch (message.messageType) {
                case messages_1.gauge.messages.Message.MessageType.SuiteDataStoreInit:
                    DataStoreFactory_1.DataStoreFactory.getSuiteDataStore().clear();
                    break;
                case messages_1.gauge.messages.Message.MessageType.SpecDataStoreInit:
                    DataStoreFactory_1.DataStoreFactory.getSpecDataStore().clear();
                    break;
                case messages_1.gauge.messages.Message.MessageType.ScenarioDataStoreInit:
                    DataStoreFactory_1.DataStoreFactory.getScenarioDataStore().clear();
                    break;
            }
            return new messages_1.gauge.messages.Message({
                messageId: message.messageId,
                messageType: messages_1.gauge.messages.Message.MessageType.ExecutionStatusResponse,
                executionStatusResponse: new messages_1.gauge.messages.ExecutionStatusResponse({
                    executionResult: new messages_1.gauge.messages.ProtoExecutionResult({
                        failed: false,
                        executionTime: 0
                    })
                })
            });
        });
    }
}
exports.DataStoreInitProcessor = DataStoreInitProcessor;
