// gauge-ts/dist/processors/ImplementationFileListProcessor.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const messages_1 = require("../gen/messages");
const Util_1 = require("../utils/Util");
class ImplementationFileListProcessor {
    process(message) {
        return __awaiter(this, void 0, void 0, function* () {
            return new messages_1.gauge.messages.Message({
                messageId: message.messageId,
                messageType: messages_1.gauge.messages.Message.MessageType.ImplementationFileListResponse,
                implementationFileListResponse: new messages_1.gauge.messages.ImplementationFileListResponse({
                    implementationFilePaths: Util_1.Util.getListOfFiles()
                })
            });
        });
    }
}
exports.ImplementationFileListProcessor = ImplementationFileListProcessor;
